// react-glide/dist/build/Glide.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Glide = void 0;

var _react = _interopRequireDefault(require("react"));

var _Preloader = require("./Preloader");

require("./reactGlide.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Glide =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Glide, _React$Component);

  function Glide() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Glide);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Glide)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "autoPlay", void 0);

    _defineProperty(_assertThisInitialized(_this), "state", {
      loading: true,
      currentIndex: 0,
      imagesLoaded: false
    });

    _defineProperty(_assertThisInitialized(_this), "startTimer", function () {
      if (_this.props.autoPlay) {
        var _this$props$autoPlayS = _this.props.autoPlaySpeed,
            _autoPlaySpeed = _this$props$autoPlayS === void 0 ? 5000 : _this$props$autoPlayS;

        _this.autoPlay = setInterval(function () {
          return _this.goToNextSlide();
        }, _autoPlaySpeed);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "goToSelectedDot", function (index) {
      _this.setState({
        currentIndex: index
      });
    });

    _defineProperty(_assertThisInitialized(_this), "goToPrevSlide", function () {
      var children = _this.props.children;
      var currentIndex = _this.state.currentIndex;
      var lastSlide = _react.default.Children.toArray(children).length - 1;
      var nextIndex = currentIndex === 0 ? lastSlide : currentIndex - 1;

      _this.setState({
        currentIndex: nextIndex
      });
    });

    _defineProperty(_assertThisInitialized(_this), "goToNextSlide", function () {
      var children = _this.props.children;
      var currentIndex = _this.state.currentIndex;
      var lastSlide = _react.default.Children.toArray(children).length - 1;
      var nextIndex = currentIndex === lastSlide ? 0 : currentIndex + 1;

      _this.setState({
        currentIndex: nextIndex
      });
    });

    return _this;
  }

  _createClass(Glide, [{
    key: "componentWillUnMount",
    value: function componentWillUnMount() {
      clearInterval(this.props.autoPlaySpeed);
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(_prevProps, prevState) {
      var prevIndex = prevState.currentIndex;
      var currentIndex = this.state.currentIndex;
      var onSlideChange = this.props.onSlideChange;

      if (currentIndex !== prevIndex && onSlideChange) {
        onSlideChange();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var currentIndex = this.state.currentIndex;
      var _this$props = this.props,
          _this$props$infinite = _this$props.infinite,
          infinite = _this$props$infinite === void 0 ? false : _this$props$infinite,
          children = _this$props.children,
          _this$props$dots = _this$props.dots,
          dots = _this$props$dots === void 0 ? true : _this$props$dots;
      var styleProps = {
        height: this.props.height,
        width: this.props.width
      };
      return _react.default.createElement("div", {
        className: "glide--container",
        style: styleProps
      }, _react.default.createElement(_Preloader.Preloader, _extends({
        startTimer: function startTimer() {
          return _this2.startTimer();
        }
      }, this.props, this.state), children), (infinite || currentIndex !== 0) && _react.default.createElement("button", {
        className: "glide--prev-btn",
        onClick: function onClick() {
          clearInterval(_this2.autoPlay);

          _this2.goToPrevSlide();
        }
      }, "\u276E"), (infinite || currentIndex !== children.length - 1) && _react.default.createElement("button", {
        className: "glide--next-btn",
        onClick: function onClick() {
          clearInterval(_this2.autoPlay);

          _this2.goToNextSlide();
        }
      }, "\u276F"), dots && _react.default.createElement("section", {
        className: "glide--dots"
      }, _react.default.Children.map(children, function (child, index) {
        return _react.default.createElement("span", {
          key: index,
          className: currentIndex === index ? 'active-dot' : 'inactive-dot',
          onClick: function onClick() {
            clearInterval(_this2.autoPlay);

            _this2.goToSelectedDot(index);
          }
        }, "\xB7");
      })));
    }
  }]);

  return Glide;
}(_react.default.Component);

exports.Glide = Glide;
