// react-glide/dist/build/Preloader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Preloader = void 0;

var _react = _interopRequireWildcard(require("react"));

var _LoadingSpinner = require("./LoadingSpinner");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Preloader =
/*#__PURE__*/
function (_Component) {
  _inherits(Preloader, _Component);

  function Preloader() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Preloader);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Preloader)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "urls", []);

    _defineProperty(_assertThisInitialized(_this), "state", {
      loading: true,
      done: false,
      loadCount: 0
    });

    _defineProperty(_assertThisInitialized(_this), "preloadImages", function () {
      var urls = _this.getImageUrls();

      if (urls.length > 0) {
        urls.map(function (src) {
          var newImage = new Image();
          newImage.src = src;
          newImage.onload = _this.updateLoadCount;
        });
      }

      if (urls.length === 0) {
        _this.updateLoadState();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "getImageUrls", function () {
      var urlResults = [];
      var children = _this.props.children;

      _react.default.Children.map(children, function (child, index) {
        var res = _this.traverseElementTree(child);

        urlResults = [].concat(_toConsumableArray(urlResults), _toConsumableArray(res));
      });

      _this.urls = _toConsumableArray(urlResults);
      return urlResults;
    });

    _defineProperty(_assertThisInitialized(_this), "traverseElementTree", function (element) {
      var results = [];

      if (element.type === 'img') {
        results.push(element.props.src);
      }

      if (element.props && element.props.children) {
        return _this.traverseElementTree(element.props.children);
      }

      return results;
    });

    _defineProperty(_assertThisInitialized(_this), "updateLoadCount", function () {
      return _this.setState(function (prevState) {
        return {
          loadCount: prevState.loadCount + 1
        };
      });
    });

    _defineProperty(_assertThisInitialized(_this), "updateLoadState", function () {
      return _this.setState(function (prevState) {
        return {
          done: true,
          loading: false
        };
      });
    });

    return _this;
  }

  _createClass(Preloader, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.preloadImages();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps, prevState) {
      var loadCountUpdated = prevState.loadCount !== this.state.loadCount;
      var allLoaded = this.state.loadCount === this.urls.length;
      var done = this.state.done !== prevState.done && this.state.done;

      if (loadCountUpdated && allLoaded) {
        this.updateLoadState();
      }

      if (done && this.props.autoPlay) {
        this.props.startTimer();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          currentIndex = _this$props.currentIndex;
      var _this$state = this.state,
          loading = _this$state.loading,
          done = _this$state.done;
      return _react.default.createElement(_react.default.Fragment, null, loading && _react.default.createElement(_LoadingSpinner.LoadingSpinner, {
        width: this.props.width
      }), done && _react.default.Children.map(children, function (child, index) {
        var className = currentIndex === index ? 'current' : '';
        return child && _react.default.createElement(child.type, _extends({
          className: "glide--item ".concat(className)
        }, child.props));
      }));
    }
  }]);

  return Preloader;
}(_react.Component);

exports.Preloader = Preloader;
